src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -871,6 +904,137 @@ ErrorIfInsertPartitionColumnDoesNotMatchSelect(Query *query, RangeTblEntry *inse
 
 
 /*
+ * FindOriginalTableId recursively traverses query tree to find actual relation
+ * id that columnExpression refers to. It returns InvolidOid if Var is a non-relational
+ * or computed value from an inner subquery or cte. This is very similar to
+ * IsPartitionColumnRecursive.
+ */
+Oid
+FindOriginalRelationId(Expr *columnExpression, List *parentQueryList, Query *query,
```

Comment:
I think I'd prefer naming this ```FindReferencedTableColumn()``` (and returning both table / column) and moving the isPartitionColumn check to the caller (if required)?

It's also not true that this is similar to IsPartitionColumnRecursive() since the latter is now implemented using this function.